### PR TITLE
Partial Armor Ignore System

### PR DESCRIPTION
**Pull request by sanek31**
-->

# Description

-->

Adds the ability to partially pierce armour.
Example:
Armor with 70% protection, when hit by a bullet with 50% penetration and 20 damage, the final damage will be 13 (as if the protection was 35%).

---

-->

- [x] Add a partial armor ignore system
- [] Some fixes
---

-->

<details><summary><h1>Media</h1></summary>
<p>

https://github.com/user-attachments/assets/d7eab844-6cca-4313-bde3-c548b1ca4f6e

</p>
</details>

---

# Changelog

-->

:cl: sanek31
- add: Added partial armor ignore system.

### CODE REVIEW

**Comment by DangerRevolution** [comment]
Why would 70% protection, with 50% piercing reduce protection to 35%?

**Comment by dge21** [comment]
because 35 is is 50% of 70

**Comment by DangerRevolution** [comment]
> because 35 is is 50% of 70

It's too early in the morning, sorry lmaoo 😭😭😭😭😭😭

**Review by DangerRevolution — APPROVED** [review]
This needs dual @VMSolidus @OldDanceJacket approval before merge

**Comment by DangerRevolution** [comment]
test fail valid it seems like?

File: Content.Shared/Damage/DamageSpecifier.cs
Comment (Mnemotechnician) [inline Content.Shared/Damage/DamageSpecifier.cs:128]: the parameter should have a meaningful name.

File: Content.Shared/Damage/DamageSpecifier.cs
Comment (Mnemotechnician) [inline Content.Shared/Damage/DamageSpecifier.cs:168]: Same as the above - those are public methods.

File: Content.Shared/Damage/Systems/DamageableSystem.cs
Comment (Mnemotechnician) [inline Content.Shared/Damage/Systems/DamageableSystem.cs:160]:
```suggestion
                    damage = DamageSpecifier.ApplyModifierSet(damage, modifierSet, bypassResistance?.Value ?? 1f);
```

File: Content.Shared/Damage/DamageSpecifier.cs
Comment (Mnemotechnician) [inline Content.Shared/Damage/DamageSpecifier.cs:153]: This is sussy and probably will never work as intended, instead making damage heal you.

**Comment by Mnemotechnician** [comment]
Also, test fails are completely legitimate. You have type errors in your code and something tells me this was not tested...

**Review by WarMechanic — COMMENTED** [review]
im not a big fan of 'armour pierce' because it ultimately means that a bullet will have a final say in whether someone takes damage. that is to say if you pitted super armour and 100% armour pierce together, armour pierce will always win regardless of how strong the armour was.

really i think we should go for new damage calculation entirely, an 'armour contest' if you will.

**Comment by sanek31** [comment]
Messed up the code. Going to rewrite it

**Comment by sanek31** [comment]
Trying to do it locally , but for some reason it's doesn't works for melee weapons.

**Comment by DangerRevolution** [comment]
> Trying to do it locally , but for some reason it's doesn't works for melee weapons.

You can find any help you need in our Discord if that's useful to you :)

**Comment by VMSolidus** [comment]
> Trying to do it locally , but for some reason it's doesn't works for melee weapons.

Melee weapons work on a different system entirely from guns, and the two have almost no overlap at all.

**Review by Pspritechologist — CHANGES_REQUESTED** [review]
I would probably suggest not taking this route for armour penetration. Generic armour penetration can be added by simply modifying DamageSpecifier to take a 'piercing' field of some kind. This field is used when applying damages and what not with an armour value coming from elsewhere. Adding the piercing value in this way allows you to do it while modifying no other functions or APIs and while automatically adding support backwards for anything that deals damage.

**Comment by sanek31** [comment]
ok

**Comment by DEATHB4DEFEAT** [comment]
> I'll close for now, I'll make another pull request when I'm done.

Reopen this one

**Pull request closed without merging**
